From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: rfresh2 <89827146+rfresh2@users.noreply.github.com>
Date: Sun, 16 Feb 2025 00:17:12 -0800
Subject: [PATCH] Handle entities with invalid position, rotation, or motion

Some world downloads have entities with NaN properties that we want to remove or reset back to valid states.

diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index b2b61203438bb1fad1ee807729781718d2467155..40b32ca68a3b689192dd7ceb593f2e8351828306 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -133,6 +133,7 @@ import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.ScoreHolder;
 import net.minecraft.world.scores.Team;
+import org.bukkit.event.entity.EntityRemoveEvent;
 import org.slf4j.Logger;
 
 public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess, ScoreHolder, ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity, ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerEntity {  // Paper - rewrite chunk system // Paper - optimise entity tracker
@@ -2534,11 +2535,25 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public void load(CompoundTag compound) {
         try {
             ListTag list = compound.getList("Pos", 6);
+            double posX = list.getDouble(0);
+            double posY = list.getDouble(1);
+            double posZ = list.getDouble(2);
+            if (archive.NanHelper.isInvalid(posX, posY, posZ)) {
+                // better to remove these than to potentially place an unbounded amount at 0,0
+                this.setRemoved(RemovalReason.DISCARDED, EntityRemoveEvent.Cause.DESPAWN);
+                return;
+            }
             ListTag list1 = compound.getList("Motion", 6);
             ListTag list2 = compound.getList("Rotation", 5);
             double _double = list1.getDouble(0);
             double _double1 = list1.getDouble(1);
             double _double2 = list1.getDouble(2);
+            _double = archive.NanHelper.ensureNonNaN(_double);
+            _double1 = archive.NanHelper.ensureNonNaN(_double1);
+            _double2 = archive.NanHelper.ensureNonNaN(_double2);
+            float yRot = archive.NanHelper.ensureNonNaN(list2.getFloat(0));
+            float xRot = archive.NanHelper.ensureNonNaN(list2.getFloat(1));
+
             this.setDeltaMovement(
                 Math.abs(_double) > 10.0 ? 0.0 : _double, Math.abs(_double1) > 10.0 ? 0.0 : _double1, Math.abs(_double2) > 10.0 ? 0.0 : _double2
             );
@@ -2549,8 +2564,8 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 Mth.clamp(list.getDouble(1), -2.0E7, 2.0E7),
                 Mth.clamp(list.getDouble(2), -3.0000512E7, 3.0000512E7)
             );
-            this.setYRot(list2.getFloat(0));
-            this.setXRot(list2.getFloat(1));
+            this.setYRot(yRot);
+            this.setXRot(xRot);
             this.setOldPosAndRot();
             this.setYHeadRot(this.getYRot());
             this.setYBodyRot(this.getYRot());
